Refactor from/to Seurat conversion

rcannood opened the pull request:
* Redesigned the from_Seurat / to_Seurat conversion functions

rcannood commented:
Note: Could use https://satijalab.github.io/seurat-object/reference/UpdateSeuratObject.html to update a Seurat <v5 object to v5 in `from_Seurat()`.

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+    graph_name <- names(graph_mapping)[[i]]
+    graph <- graph_mapping[[i]]
+    if (!is.character(graph) || length(graph) != 1) {
+      stop("each graph must be a character vector of length 1")
```
Maybe this is clearer:

```suggestion
      stop("item ", graph_name, "in graph_mapping is not a character vector of length 1")
```

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
-    seurat_obj[[key]] <- SeuratObject::CreateAssayObject(counts = layer_)
+  # trackstatus: class=Seurat, feature=get_uns, status=done
+  # trackstatus: class=Seurat, feature=get_varp, status=done
+  obj@misc <- list()
```
Can probably also avoid the `@` here

LouiseDck replied:
Out of curiosity, why do you want to avoid the `@`?

rcannood replied:
I believe @ is usually used for accessing internal data structures which end users should not be messing around with. 

I'm not sure whether there is always an easy alternative. Let me know if you find one :)

rcannood replied:
Doesn't work as a replacement for this line of code.

```
SeuratObject::Misc(obj) <- list()
Error in `Misc<-.Seurat`(`*tmp*`, value = list()) : 
  argument "slot" is missing, with no default
```

In this case, we could **assume** that `obj@misc` is an empty list to begin with.

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+        paste0("'", expected_slots, "'", collapse = ", ")
+      ))
+    }
+    misc_data <- adata[[misc[[1]]]]
```
Is it possible to get the names without getting everything in the slot? Just because for `H5ADAnnData` this will drag everything in that slot into memory (I think).

rcannood replied:
Good point. Probably not, unless we 

```r
if (slot_name == "obs") adata$obs_keys()
elif (slot_name == "var") adata$var_keys()
elif ...
```

lazappi replied:
In theory yes (or a case statement might be nicer) but I think at the moment `adata$obs_keys()` etc. read everything anyway so not much point unless those are changed. Maybe don't worry about it for this PR.

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+      stop("misc_mapping must be a named list with one or two elements")
+    }
+    expected_slots <- c("X", "layers", "obs", "obsm", "obsp", "var", "varm", "varp", "uns")
+    if (!misc[[1]] %in% expected_slots) {
```
The code would be a bit easier to follow if there was something like:

```r
misc_slot <- misc[[1]]
misc_key <- misc[[2]]
```

rcannood replied:
done!

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
   }
 
-  names
+  if (!.to_seurat_is_atomic_character(layer_name)) {
+    stop("layer_name must be the name of one of the layers or NULL")
+  }
+
+  if (!layer_name %in% names(adata$layers)) {
+    stop("layer_name must be the name of one of the layers or NULL")
+  }
```
These could have different messages to make it clearer what the problem is.

rcannood replied:
true!

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+}
+
+.to_seurat_get_matrix <- function(adata, layer_name) {
+  if (is.null(layer_name)) {
+    return(Matrix::t(adata$X))
+  }
```
Do you need to do this or can you assume there will always be a name? It might depend what the guessing function do which I haven't looked at yet.

rcannood replied:
well you need to have some way of allowing the user to use `$X` as the seurat counts of data, so by using `NULL` the `$X` will be used.

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+  }
+  embed <- adata$obsm[[obsm_embedding]]
+
+  if (is.null(embed)) {
+    warning(paste0("The reduction ", obsm_embedding, " is not present in adata$obsm, skipping"))
+    return(NULL)
+  }
```
I think other missing things are an error rather than a warning?

rcannood replied:
True. This was a leftover from when the defaults of to_seurat and from_seurat just had a bunch of defaults without checking whether these exist in the original anndata / seurat object.

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+    args$loadings <- loadings
+  }
+
+  do.call(SeuratObject::CreateDimReducObject, args)
```
Is there a reason to do it this way rather than just calling the function with arguments?

rcannood replied:
fixed!

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+
+  layers <- list()
+
+  if (!is.null(adata$X)) {
+    layers["counts"] <- list(NULL)
+  }
```
Ah, I guess you use `NULL` to mean `adata$X` rather than `adata$layers["X"]`. I missed that before.

rcannood replied:
:+1:

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+#' @export
+#'
+#' @rdname to_Seurat
 #' @examples
 #' ad <- AnnData(
 #'   X = matrix(1:5, 3L, 5L),
 #'   obs = data.frame(row.names = LETTERS[1:3], cell = 1:3),
 #'   var = data.frame(row.names = letters[1:5], gene = 1:5)
 #' )
 #' to_Seurat(ad)
```
It would be nice to have some examples showing how the mapping works.

rcannood replied:
true!

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+#'
+#' If `NULL`, the internal function `from_Seurat_guess_layers` will be used to guess the layer mapping as follows:
+#'
+#' * All Seurat layers are copied by name.
```
I assume to `adata$layers` and `adata$X` is empty?

rcannood replied:
Yes, I'll clarify.

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+#' A named list to map Seurat reductions to AnnData `obsm`. Each item in the list must be a named list with keys 'key',
+#' 'obsm', and 'varm'.
+#'
+#' Example: `obsm_mapping = list(pca = list(key = "PC_", obsm = "X_pca", varm = "PCs"))`.
```
I'll see below but it should probably be possible for some of these to be missing

rcannood replied:
Sorry, the documentation is no longer correct, I'll fix.

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+#'
+#' * All Seurat reductions are prefixed with `X_` and copied to AnnData `obsm`.
+#'
+#' @section Varm mapping:
+#'
+#' A named list to map Seurat PCA loadings to AnnData `varm`.
```
Is `obsm_mapping` also copying things to `varm`? Do we need both?

rcannood replied:
Sorry, the `obsm_mapping` documentation was wrong -- is fixed now.

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+#'
+#' If `NULL`, the internal function `from_Seurat_guess_varms` will be used to guess the varm mapping as follows:
+#'
+#' * The name of the PCA loadings is copied by name.
```
Only PCA or any other reduction that has loadings?

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+#'
+#' * The name of the PCA loadings is copied by name.
+#'
+#' @section Obsp mapping:
```
I would not capitalise obsp and maybe format it as code

rcannood replied:
Good point, still need to change this

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+#'
+#' @section Varp mapping:
+#'
+#' A named list to map Seurat miscellaneous data to AnnData `varp`. Each item in the list must be a character
+#' vector of length 3 or less. The first element must be `"misc"`. The second element is the name of the data in the
+#' corresponding slot. The third element is used to subset the data even further.
+#'
+#' Example: `varp_mapping = list(foo = c("misc", "foo"))`.
+#'
+#' If `NULL`, the internal function `from_Seurat_guess_varps` will be used to guess the varp mapping as follows:
+#'
+#' * No data is mapped to `varp`.
```
This bit I don't understand. Why have a first element if it's always the same and what does the third element do?

rcannood replied:
Tried to clarify the docs

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+#' obj <- RunUMAP(obj, dims = 1:10)
+#' from_Seurat(obj)
 # nolint start: object_name_linter
 from_Seurat <- function(
```
Maybe these functions should be called something like `anndata_from_seurat()` just to make it a bit clearer?

LouiseDck replied:
I disagree, as we only aim to convert from & to anndata.
However, if we want to emphasize this, I would prefer `anndata_to_seurat()` and `seurat_to_anndata()`

rcannood replied:
I'd be fine with:

* `anndata_to_seurat` & `seurat_to_anndata`
* `to_seurat` & `from_seurat`
* or keep it as is, namely `to_Seurat` and `from_Seurat`

rcannood replied:
It's possible to expose `ad$to_Seurat()` and not `to_Seurat(ad)`, but we'll need to expose `from_Seurat(obj)` as this functionality is otherwise not accessible. This is the status quo ^^

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
     output_class = c("InMemoryAnnData", "HDF5AnnData"),
-    assay = NULL,
-    X = "counts",
+    assay_name = "RNA",
```
This could be the name of the first assay by default in case it's something else

rcannood replied:
true -- `obj@active.assay`?

rcannood replied:
I changed the default to NULL. If set to NULL, `SeuratObject::DefaultAssay(seurat_obj)` will be used.

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
-    assay_name <- assay
-  } else {
-    assay_name <- SeuratObject::DefaultAssay(seurat_obj)
+  if (is.null(seurat_assay) || !inherits(seurat_assay, "Assay5")) {
+    stop(paste0("Assay '", assay_name, "' is not a valid Seurat v5 assay"))
+  }
```
I think this is checking two things. I would probably have one check for the object version (with a prompt to upgrade it) and a second check later for the assay name.

rcannood replied:
I'll add it to the error message!

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+  }
 
-    # If Seurat object contains multiple assays, notify user the Default one is used
-    if (length(names(seurat_obj@assays)) > 1) {
-      message(
-        "There are ", length(names(seurat_obj@assays)), " assays in the Seurat object; using the default assay ('",
-        assay_name, "'). You can use the `assay` parameter to select a specific assay."
-      )
-    }
```
I think this is nicer behaviour, I would put this back

LouiseDck replied:
I agree, if you want to pick a default assay to convert it should not be `RNA` by default, but the `DefaultAssay`

rcannood replied:
Fixed!

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+  # fetch obs
   # trackstatus: class=Seurat, feature=set_obs_names, status=done
   # trackstatus: class=Seurat, feature=set_obs, status=done
   obs <- seurat_obj@meta.data
```
I'm pretty sure `seurat_obj[[]]` will get this but would need to double check

rcannood replied:
Ok, let me know ^^

rcannood replied:
[`@meta.data`](https://satijalab.github.io/seurat-object/reference/Seurat-class.html) is a documented slot of the Seurat -- I like the more explicit nature of this notation.

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
   # use generator to create new AnnData object
   generator <- get_anndata_constructor(output_class)
-  ad <- generator$new(
+  adata <- generator$new(
     obs = obs,
     var = var,
     ...
   )
```
Maybe it's better to wait until the end of the function to create this? Otherwise you could end up with incomplete files for `H5ADAnnData`.

rcannood replied:
True. Otoh, it would make the whole operation a whole lot less memory efficient.

I could add a tryCatch which removes the file on error. WDYT?

lazappi replied:
Maybe, if there's a nice way to do it. The way we write stuff at the moment is a bit risky in general so maybe it's more a design flaw in `H5ADAnnData`...

rcannood replied:
I changed it to:

```

  tryCatch(
    {
      generator$new(
        obs = obs,
        var = var,
        ...
      )
      # ... add more data ...
    },
    error = function(e) {
      if (output_class == "HDF5AnnData") {
        on.exit(cleanup_HDF5AnnData(...))
      }
      stop(e)
    }
  )
```

wdyt?

lazappi reviewed (changes requested):
I made a bunch of comments, most of them are fairly minor. I haven't tried running anything yet though so possible I missed things.

LouiseDck commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+  for (i in seq_along(layer_mapping)) {
+    from <- layer_mapping[[i]]
+    to <- names(layer_mapping)[[i]]
+    if (to == "counts" || to == "data") {
+      next
+    }
```
I don't mind this control flow

rcannood replied:
I changed 

```r
    if (to == "counts" || to == "data") {
      next
    }
```

to 

```r
    if (!to %in% c("counts", "data")) {
      SeuratObject::LayerData(obj, assay = assay_name, layer = to) <- adata$layers[[from]]
    }
```

LouiseDck commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+  # trackstatus: class=Seurat, feature=get_obsm, status=wip
+  # trackstatus: class=Seurat, feature=get_varm, status=wip
+  if (!is.null(reduction_mapping)) {
+    if (!is.list(reduction_mapping) || (length(reduction_mapping) > 0 && is.null(names(reduction_mapping)))) {
+      stop("reduction_mapping must be a named list")
     }
```
Lots of checking, but it is still possible that not all elements of the list are named. If e.g. only one of the elements is named, the unnamed reductions get mapped to an "X" reduction in Seurat, and overwrite each other.

rcannood replied:
To not make it too complicated, can we assume that reduction_mapping is always a named list?

LouiseDck commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
-        x = integer(0),
-        dims = c(obj$n_vars(), obj$n_obs())
+      if (
+        !is.list(reduction) || is.null(names(reduction)) ||
+          !all(names(reduction) %in% c("key", "obsm", "varm")) ||
+          !all(c("key", "obsm", "varm") %in% names(reduction))
```
I have an issue with requiring the "varm" name in the list: for many dimensionality reductions you do not get loadings in "varm", so requiring it seems a bit harsh?

rcannood replied:
+1!

LouiseDck commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
 #'
+#' @section Layer mapping:
+#'
+#' A named list to map AnnData layers to Seurat layers. Each item in the list must be a character vector of length 1.
+#' The `$X` key maps to the `X` slot.
+#'
+#' Example: `layer_mapping = list(counts = "counts", "$X" = "data", foo = "bar")`.
+#'
+#' If `NULL`, the internal function `from_Seurat_guess_layers` will be used to guess the layer mapping as follows:
+#'
+#' * All Seurat layers are copied by name.
```
I want it emphasized that this means that by default, `adata$X` will be `NULL`, and that if you want to prevent this you should prevent a mapping.

rcannood replied:
+1!

LouiseDck commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+#'
+#' A named list to map Seurat miscellaneous data to AnnData `varp`. Each item in the list must be a character
+#' vector of length 3 or less. The first element must be `"misc"`. The second element is the name of the data in the
+#' corresponding slot. The third element is used to subset the data even further.
```
Maybe this is my inexperience with Seurat showing, but what do you mean by `used to subset the data further`?

rcannood replied:
Phrasing was improved

LouiseDck commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+
+  # fetch obsm
+  # trackstatus: class=Seurat, feature=set_obsm, status=wip
+  for (i in seq_along(obsm_mapping)) {
+    obsm <- obsm_mapping[[i]]
+    obsm_name <- names(obsm_mapping)[[i]]
+
+    if (!is.character(obsm) || length(obsm) != 2) {
+      stop("each obsm_mapping must be a character vector of length 2")
+    }
+
+    key1 <- obsm[[1]]
+    key2 <- obsm[[2]]
+
+    if (key1 == "reductions") {
+      adata$obsm[[obsm_name]] <- seurat_obj@reductions[[key2]]@cell.embeddings
+    } else if (key1 == "misc") {
+      adata$obsm[[obsm_name]] <- seurat_obj@misc[[key2]]
+    }
+  }
```
I might overlook this, but there's no need to specifiy "varm" in the mapping, right? In that case, I think the documentation of the function needs an update.

rcannood replied:
not sure what you mean -- could you clarify?

LouiseDck replied:
Yes -- My comment was wrongly placed and you already solved part of it :tada:

LouiseDck commented on DESCRIPTION:
```diff
@@ -65,12 +65,12 @@ Imports:
     R6
 Suggests:
     anndata,
-    BiocStyle,
```
cf #169 but we should indeed discuss this

rcannood replied:
Removed changes that were not related to Seurat and moved the discussion to #191 !

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+#' the mapping.
+#'
+#' @param adata An AnnData object to be converted
+#' @param assay_name Name of the assay to be created
+#' @param layer_mapping A named list to map AnnData layers to Seurat layers. See section "Layer mapping" for more
+#'   details.
+#' @param reduction_mapping A named list to map AnnData reductions to Seurat reductions. Each item in the list must be a
+#'   named list with keys 'key', 'obsm', and 'varm'. See section "Reduction mapping" for more details.
+#' @param graph_mapping A named list to map AnnData graphs to Seurat graphs. Each item in the list must be a character
+#'   vector of length 1. See section "Graph mapping" for more details.
+#' @param misc_mapping A named list to map miscellaneous data to the names of the data in the Seurat object. See section
+#'   "Miscellaneous mapping" for more details.
```
At the moment I named the arguments based on what the object that ends up being created. In `to_Seurat` we're deciding on what needs to be copied into the Seurat object, which is why the argument names are named after seurat object slots.

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+#' @param misc_mapping A named list to map miscellaneous data to the names of the data in the Seurat object. See section
+#'   "Miscellaneous mapping" for more details.
+#'
+#' @section Layer mapping:
+#'
+#' A named list to map AnnData layers to Seurat layers. Each item in the list must be a character vector of length 1.
+#' A value of `NULL` corresponds to the `X` slot.
+#'
+#' Example: `layer_mapping = list(counts = "counts", data = NULL, foo = "bar")`.
```
Fixed!

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+#'
+#' Example: `layer_mapping = list(counts = "counts", data = NULL, foo = "bar")`.
+#'
+#' If `NULL`, the internal function `to_Seurat_guess_layers` will be used to guess the layer mapping as follows:
```
Originally each of the functions were being exported but this resulted in a lot of clutter in the package documentation, which is why they are currently internal functions. Would you prefer to export them?

lazappi replied:
Not necessarily, pros and cons. But if they were exported then the documentation should go there.

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+#'
+#' @section Reduction mapping:
+#'
+#' A named list to map AnnData `obsm` and `varm` to Seurat reductions. Each item in the list must be a named list
```
Hmm. Or what about "AnnData `$obsm` and `$varm`"?

lazappi replied:
Yeah, maybe. I just think it's easy to get mixed up between different objects/arguments so it helps to be as clear as possible. I don't really mind how exactly, as long as it's consistent.

rcannood replied:
Tried to make the notations more consistent!

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+#'
+#' @section Graph mapping:
+#'
+#' A named list mapping graph names to the names of the graphs in the AnnData object. Each item in the list must be a
+#' character vector of length 1.
```
Fixed!

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
 #'
-#' @param obj An AnnData object
+#' A named list mapping miscellaneous data to the names of the data in the AnnData object. Each item in the list must be
+#' a named list with one or two elements. The first element must be one of: 'X', 'layers', 'obs', 'obsm', 'obsp', 'var',
```
Fixed!

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
-#' @param obj An AnnData object
+#' A named list mapping miscellaneous data to the names of the data in the AnnData object. Each item in the list must be
+#' a named list with one or two elements. The first element must be one of: 'X', 'layers', 'obs', 'obsm', 'obsp', 'var',
+#' 'varm', 'varp', 'uns'. The second element is the name of the data in the corresponding slot. If the second element is
+#' not present, the data in the first element will be used.
```
I tried to clarify -- is this better?

lazappi replied:
Yep, as long as it matches what it does in practice. What it says now is different to what I understood before.

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+#' If `NULL`, the internal function `to_Seurat_guess_misc` will be used to guess the miscellaneous mapping as follows:
+#'
+#' * If `adata$obsp` is defined, this is mapped to a misc value named `obsp`.
+#' * If `adata$uns` is defined, this is mapped to a misc value named `uns`.
```
:+1:!

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+    graph_mapping = NULL,
+    misc_mapping = NULL) {
+  # nolint end: object_name_linter
   requireNamespace("SeuratObject")
```
:+1:!

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+  var_names <- adata$var_names[]
+
+  # check seurat layers
+  if (is.null(names(layer_mapping))) {
+    names(layer_mapping) <- layer_mapping
+  }
```
I'd say it's okay with using `layer_mapping = c("counts", "data")` as a synonym for `c(counts = "counts", data = "data")`.

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+
+  # trackstatus: class=Seurat, feature=get_var, status=done
+  if (!is.null(adata$var)) {
+    obj@assays[[assay_name]] <- SeuratObject::AddMetaData(
```
Let me know if you'd prefer something else :relaxed:

rcannood replied:
`GetAssayData` and `SetAssayData` returns the data of the assay, not the assay itself.

I can use `obj[[assay_name]]` instead of `obj@assays[[assay_name]]` but I don't see the added benefit. If anything, `obj[[assay_name]]`  allows me to accidentally interact with data that is not actually an assay.

Upon looking closer, I see that SeuratObject provides [documentation of the Seurat class](https://satijalab.github.io/seurat-object/reference/Seurat-class.html) for slots like `@assays`, `@graphs`, ..., so I'd say using these slots is actually ok.

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+  # make sure obs and var names are set properly
+  # trackstatus: class=Seurat, feature=get_obs_names, status=done
+  # trackstatus: class=Seurat, feature=get_var_names, status=done
+  colnames(obj) <- obs_names
+  rownames(obj) <- var_names
```
Maybe. Seurat might have done something with those dimnames by then...

lazappi replied:
Ok. If there is a reason for it then it's fine.

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+  # trackstatus: class=Seurat, feature=get_obsm, status=wip
+  # trackstatus: class=Seurat, feature=get_varm, status=wip
+  if (!is.null(reduction_mapping)) {
+    if (!is.list(reduction_mapping) || (length(reduction_mapping) > 0 && is.null(names(reduction_mapping)))) {
+      stop("reduction_mapping must be a named list")
     }
```
:+1:

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,815 @@
+      stop("each obsm_mapping must be a character vector of length 2")
+    }
+
+    key1 <- obsm[[1]]
+    key2 <- obsm[[2]]
```
+1

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,830 @@
+  if ("pca" %in% names(seurat_obj@reductions)) {
+    # Check if the dimreduc was calculated by the selected assay
+    reduction <- seurat_obj@reductions[["pca"]]
+    if (reduction@assay.used == assay_name) {
```
I can't find a better way to get the assay used so probably have to use `@` here

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,830 @@
+
+  obsp_mapping <- list()
+
+  for (graph_name in names(seurat_obj@graphs)) {
```
```suggestion
  for (graph_name in SeuratObject::Graphs(seurat_obj)) {
```

lazappi commented on R/Seurat.R:
```diff
@@ -1,205 +1,830 @@
+  obsp_mapping <- list()
+
+  for (graph_name in names(seurat_obj@graphs)) {
+    graph <- seurat_obj@graphs[[graph_name]]
```
```suggestion
    graph <- seurat_obj[[graph_name]]
```

rcannood replied:
`@graphs` is kept as a precaution

lazappi reviewed (changes requested):
Mostly adding suggestions for accessing stuff from a Seurat object. I didn't comment on every line but I think there should be an example for each slot.

LouiseDck commented on R/Seurat.R:
```diff
@@ -1,205 +1,830 @@
+#' @param assay_name The name of the assay to be converted. If `NULL`, the default assay will be used.
+#' @param layer_mapping A named list mapping layer names to the names of the layers in the Seurat object. Each item in
+#' the list must be a character vector of length 1. See section "Layer mapping" for more details.
+#' @param obsm_mapping A named list mapping reduction names to the names of the reductions in the Seurat object.
+#' Each item in the list must be a named list with keys 'key', 'obsm', and 'varm'. See section "Reduction mapping" for
```
This seems wrong -- it also does not correspond to the "obsm mapping" section.

rcannood replied:
Thanks for noticing, has been fixed!

rcannood commented on R/check_requires.R:
```diff
@@ -0,0 +1,34 @@
+      c(
+        "{what} requires the {.pkg {missing}} package{?s}",
+        "i" = paste(
+          "To continue, install {cli::qty(missing)}{?it/them} using",
+          "{.code install.packages(c({missing_str}))}"
```
> These instructions won't work for Bioconductor packages.

Fixed!

> updating all messages/warning/errors to use {cli} if we are going to depend on it anyway?

I'd like that, but let's do this in a separate PR :sweat_smile:

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,830 @@
-      stop("X must be NULL or one of: 'counts', 'data', 'scale.data'")
-    }
+  if (is.null(assay_name)) {
+    assay_name <- SeuratObject::DefaultAssay(seurat_obj)
```
This would bypass the `requiredNamespace("SeuratObject")` ^^

I'll make a note in the docs.

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,830 @@
-    assay_name <- assay
-  } else {
-    assay_name <- SeuratObject::DefaultAssay(seurat_obj)
+  seurat_assay <- seurat_obj@assays[[assay_name]]
```
`GetAssayData()` returns the data layer of the assay, not the assay object itself.

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,830 @@
+    key2 <- obsp[[2]]
+
+    if (key1 == "graphs") {
+      adata$obsp[[obsp_name]] <- as(seurat_obj@graphs[[key2]], "sparseMatrix")
```
Then lets stick to using @graphs, as this is more explicit. I don't want strange errors to occur because a user accidentally provided the name of a DR as a graph name.

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,830 @@
+
+  layers_mapping <- list()
+
+  for (layer_name in names(seurat_assay@layers)) {
```
The code was used to use these functions where possible :)

rcannood commented on R/Seurat.R:
```diff
@@ -1,205 +1,830 @@
+
+  for (reduction_name in names(seurat_obj@reductions)) {
+    # Check if the dimreduc was calculated by the selected assay
+    reduction <- seurat_obj@reductions[[reduction_name]]
```
@reductions is kept as a safety precaution

rcannood commented:
Phew. I think I addressed most of the comments. 

* Wherever SeuratObject had functions for accessing certain data, these functions were used (e.g. changed `names(obj@layers)` for `SeuratObject::Layers(obj)`). However, I didn't change code when the alternative was less specific (e.g. changing `obj@assays[[assay_name]]` for `obj[[assay_name]]`) or readable (e.g. changing `obj@meta.data` for `obj[[]]`) to avoid accidental bugs.
* Tried to clean up with `cleanup_HDF5AnnData` when an error occurs while converting to HDF5AnnData. @lazappi Can you take a look?
* Changing all `stop()`, `warning()`, `cat()`, etc.. to use **{cli}** sounds great, but let's do that in a separate PR

lazappi commented on R/SingleCellExperiment.R:
```diff
@@ -168,11 +171,19 @@ from_SingleCellExperiment <- function(
+      ...
+    ),
+    error = function(e) {
+      if (output_class == "HDF5AnnData") {
+        on.exit(cleanup_HDF5AnnData(...))
+      }
```
Should this be outside the error part so the cleanup is still run if one of the later steps fails?

rcannood replied:
I think there is no code after this call in the function. The SCE converter is simply a bit limited at this stage

lazappi reviewed (approved):
I think I'm happy with this now
